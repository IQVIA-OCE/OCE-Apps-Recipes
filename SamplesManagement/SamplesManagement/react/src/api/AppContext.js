import api from '../utils/api';

export const fetchUser = userId =>
  api.query(
    `SELECT Id, Name, OCE__ProfileId__c FROM User WHERE Id = '${userId}'`
  );