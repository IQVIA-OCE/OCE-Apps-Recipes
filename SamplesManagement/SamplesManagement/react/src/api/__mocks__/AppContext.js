export const fetchUser = jest.fn().mockResolvedValue([
  [
    {
      Id: '0050k000004CineAAC',
      Name: 'SALESREP',
      OCE__ProfileId__c: '00e0k000000O1zxAAC',
    },
  ],
  {
    totalSize: 1,
    done: true,
  },
]);
