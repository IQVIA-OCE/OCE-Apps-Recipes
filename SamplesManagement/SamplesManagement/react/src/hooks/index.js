export { useFetcher } from './useFetcher';
export { useBoolean } from './useBoolean';
export { useHandleData } from './useHandleData';
export { useBanner } from './useBanner';
