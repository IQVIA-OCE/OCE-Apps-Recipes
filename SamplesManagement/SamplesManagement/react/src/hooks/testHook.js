import renderer from 'react-test-renderer';
import React from 'react';

const TestHook = ({ callback }) => {
  callback();
  return null;
};

export const testHook = callback =>
  renderer.create(<TestHook callback={callback} />);

const TestRenderHook = ({ callback }) => {
  return callback();
};

export const testRenderHook = callback =>
  renderer.create(<TestRenderHook callback={callback} />);
