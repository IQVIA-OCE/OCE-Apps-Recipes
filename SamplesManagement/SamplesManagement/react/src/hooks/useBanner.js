import { useEffect, useState } from 'react';

export const useBanner = () => {
  const [banner, setBanner] = useState({
    variant: '',
    message: '',
    visible: false,
    icon: '',
  });

  useEffect(() => {
    const timer = setTimeout(
      () => setBanner(prev => ({ ...prev, visible: false })),
      4000
    );

    return () => clearTimeout(timer);
  }, [banner.visible]);

  return [banner, setBanner];
};
