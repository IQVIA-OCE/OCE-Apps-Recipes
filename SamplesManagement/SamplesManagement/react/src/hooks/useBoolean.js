import { useState } from 'react';

export const useBoolean = initialValue => {
  const [value, setValue] = useState(initialValue);

  const toggle = () => setValue(v => !v);
  const setTrue = () => setValue(true);
  const setFalse = () => setValue(false);

  return [value, { toggle, setTrue, setFalse }];
};
