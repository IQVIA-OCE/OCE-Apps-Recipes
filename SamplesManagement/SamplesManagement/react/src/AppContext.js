import React, { Component } from 'react';
import { sfNetAPI, environment } from '@oce-apps/oce-apps-bridges';
import { View } from 'react-native';
import { ActivityIndicator, withTheme } from '@oce-apps/apollo-react-native';
import { fetchUser } from './api/AppContext';
import { NAMESPACE } from './constants/constants';

const defaultContextValue = {
  isLoading: true,
  userId: null,
  profileId: null,
  orgId: null,
  error: null,
};

export const AppContext = React.createContext(defaultContextValue);

class AppContextProvider extends Component {
  state = defaultContextValue;

  getInitialData = async () => {
    try {
      const userId = environment.userId();
      const [[user]] = await fetchUser(userId);

      this.setState({
        profileId: user[`${NAMESPACE}ProfileId__c`],
        username: user.Name,
        isLoading: false,
        error: null,
        userId
      });
    } catch (error) {
      this.setState({ isLoading: false, error });
    }
  };

  componentDidMount() {
    this.getInitialData();
  }

  render() {
    const { theme } = this.props;

    return (
      <AppContext.Provider value={this.state}>
        {this.state.isLoading ? (
          <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center', backgroundColor: theme.colors.background }}>
            <ActivityIndicator
              animating={true}
              color={theme.colors.primary}
              style={{ paddingVertical: 10 }}
            />
          </View>
        ) : (
          this.props.children
        )}
      </AppContext.Provider>
    );
  }
}

export default withTheme(AppContextProvider);
