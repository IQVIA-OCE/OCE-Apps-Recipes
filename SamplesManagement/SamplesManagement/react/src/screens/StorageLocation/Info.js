import React, { useContext } from 'react';
import { StyleSheet, View } from 'react-native';
import InfoField from './InfoField';
import { AppContext } from '../../AppContext';
import { useFetcher, useHandleData } from '../../hooks';
import { fetchUsers } from '../../api/StorageLocation';
import { normalizeUsers } from './utils';

const Info = ({ values }) => {
  const { username, userId } = useContext(AppContext);
  const ids = [userId];

  if (values.createdById !== userId) {
    ids.push(values.createdById);
  }
  if (values.modifiedById !== userId) {
    ids.push(values.modifiedById);
  }

  const [users] = useFetcher(async () => await fetchUsers(ids), normalizeUsers);

  return useHandleData(users)(data => (
    <View style={styles.form}>
      <View style={styles.col}>
        <InfoField
          label="Created By"
          text={`${
            data[values.createdById] ? data[values.createdById] : username
          } ${values.createdDate}`}
        />
      </View>
      <View style={styles.col}>
        <InfoField
          label="Last Modified By"
          text={`${
            data[values.modifiedById] ? data[values.modifiedById] : username
          } ${values.modifiedDate}`}
        />
      </View>
    </View>
  ));
};

const styles = StyleSheet.create({
  form: {
    padding: 10,
    flexDirection: 'row',
  },
  col: {
    paddingHorizontal: 15,
    flex: 1,
  },
});

export default Info;
