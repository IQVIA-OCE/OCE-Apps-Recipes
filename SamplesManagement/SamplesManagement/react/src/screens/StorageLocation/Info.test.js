import React from 'react';
import renderer from 'react-test-renderer';
import Info from './Info';
import { AppContext } from '../../../AppContext';
import { useFetcher, useHandleData } from '../../hooks';

jest.mock('../../hooks/index');

const context = {
  username: 'test user name',
  userId: '1',
};

describe('StorageLocation Info', () => {
  beforeAll(() => {
    useFetcher.mockImplementation((fetch, normalizer) => {
      console.log(fetch, normalizer)
      return [
        {
          loading: false,
          data: {
            '2': 'Second user',
          },
        },
      ]
    });
    useHandleData.mockImplementation(({ data }) => fn => fn(data));
  });

  it('Should render Info component user data', async () => {
    const tree = renderer.create(
      <AppContext.Provider value={context}>
        <Info
          values={{
            createdById: '1',
            modifiedById: '1',
            createdDate: 'created date',
            modifiedDate: 'modified date',
          }}
        />
      </AppContext.Provider>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('Should render Info component other data', async () => {
    const tree = renderer.create(
      <AppContext.Provider value={context}>
        <Info
          values={{
            createdById: '2',
            modifiedById: '2',
            createdDate: 'created date',
            modifiedDate: 'modified date',
          }}
        />
      </AppContext.Provider>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
