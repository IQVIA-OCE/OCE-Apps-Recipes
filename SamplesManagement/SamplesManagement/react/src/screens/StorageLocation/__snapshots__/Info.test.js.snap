// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorageLocation Info Should render Info component other data 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": 15,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 15,
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "marginBottom": 15,
            },
          ]
        }
      >
        Created By
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "fontSize": 16,
            },
          ]
        }
      >
        Second user created date
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": 15,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 15,
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "marginBottom": 15,
            },
          ]
        }
      >
        Last Modified By
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "fontSize": 16,
            },
          ]
        }
      >
        Second user modified date
      </Text>
    </View>
  </View>
</View>
`;

exports[`StorageLocation Info Should render Info component user data 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": 15,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 15,
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "marginBottom": 15,
            },
          ]
        }
      >
        Created By
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "fontSize": 16,
            },
          ]
        }
      >
        test user name created date
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": 15,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 15,
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "marginBottom": 15,
            },
          ]
        }
      >
        Last Modified By
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "fontSize": 16,
            },
          ]
        }
      >
        test user name modified date
      </Text>
    </View>
  </View>
</View>
`;
